sklearn/datasets/_openml.py

```diff
@@ -412,64 +412,98 @@ def _get_num_samples(data_qualities: OpenmlQualitiesType) -> int:
-            for line in response:
-                actual_md5_checksum.update(line)
-                yield line.decode("utf-8")
+    data_home : str
+        The location where to cache the data.
 
-        stream = _stream_checksum_generator(response)
+    parser : {"liac-arff", "pandas"}
+        The parser used to parse the ARFF file.
 
-        encode_nominal = not output_arrays_type == "pandas"
-        return_type = _arff.COO if output_arrays_type == "sparse" else _arff.DENSE_GEN
+    output_type : {"numpy", "pandas", "sparse"}
+        The type of the arrays that will be returned. The possibilities ara:
 
-        arff = _arff.load(
-            stream, return_type=return_type, encode_nominal=encode_nominal
-        )
+        - `"numpy"`: both `X` and `y` will be NumPy arrays;
+        - `"sparse"`: `X` will be sparse matrix and `y` will be a NumPy array;
+        - `"pandas"`: `X` will be a pandas DataFrame and `y` will be either a
+          pandas Series or DataFrame.
 
-        X, y, frame, nominal_attributes = _liac_arff_parser(
-            arff,
-            output_arrays_type,
-            features_dict,
-            data_columns,
-            target_columns,
-            col_slice_x,
-            col_slice_y,
-            shape,
-        )
+    openml_columns_info : dict
+        The information provided by OpenML regarding the columns of the ARFF
+        file.
 
-        # consume remaining stream, if early exited
-        for _ in stream:
-            pass
+    feature_names_to_select : list of str
+        The list of the features to be selected.
 
-        if actual_md5_checksum.hexdigest() != md5_checksum:
-            raise ValueError(
-                "md5 checksum of local file for "
-                + url
-                + " does not match description. "
-                "Downloaded file could have been modified / "
-                "corrupted, clean cache and retry..."
-            )
+    target_names_to_select : list of str
+        The list of the target variables to be selected.
+
+    shape : tuple or None
+        With `parser="liac-arff"`, when using a generator to load the data,
+        one needs to provide the shape of the data beforehand.
 
-        return X, y, frame, nominal_attributes
+    md5_checksum : str
+        The MD5 checksum provided by OpenML to check the data integrity.
+
+    Returns
+    -------
+    X : {ndarray, sparse matrix, dataframe}
+        The data matrix.
+
+    y : {ndarray, dataframe, series}
+        The target.
+
+    frame : dataframe or None
+        A dataframe containing both `X` and `y`. `None` if
+        `output_array_type != "pandas"`.
+
+    categories : list of str or None
+        The names of the features that are categorical. `None` if
+        `output_array_type == "pandas"`.
+    """
+    gzip_file = _open_openml_url(url, data_home, n_retries=n_retries, delay=delay)
+    actual_md5_checksum = hashlib.md5(gzip_file.read()).hexdigest()
+    gzip_file.close()
```

Comment:
This new implementation reads the file once for MD5 checksum and another time for loading the data.

Can we use the same idea in `main` where we have a `_stream_checksum_generator` and pass the stream to `load_arff_from_gzip_file`?

https://github.com/scikit-learn/scikit-learn/blob/aee564c544e3245e52bd413709e43f192ad02ba9/sklearn/datasets/_openml.py#L434-L437

Reply:
It is not a good idea. I implemented it at the beginning of the PR. It was required to make our own `GZipFile` class and dig into the internals of pandas to know how to handle this case (done with @jorisvandenbossche). Once done, we compare the timing on a huge file like MNIST and the complexity of the solution was not worth it.